### PR TITLE
CSHARP-3086: Hidden Indexes.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG: https://evergreen.mongodb.com/version/5f5f7e0b0305b97395e128f1

### CODE REVIEW

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexRequestTests.cs [hunk 1]
```diff
@@ -689,6 +711,20 @@ public void CreateIndexDocument_should_throw_when_Collation__is_set_and_is_not_s
             result.Should().Be(value);
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void Hidden_get_and_set_should_work(
+            [Values(null, false, true)] bool? valueString)
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexRequestTests.cs]: Rename parameter to `value`

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexRequestTests.cs [hunk 2]
```diff
@@ -689,6 +711,20 @@ public void CreateIndexDocument_should_throw_when_Collation__is_set_and_is_not_s
...
+        {
+            var subject = new CreateIndexRequest(new BsonDocument("x", 1));
+            var value = valueString == null ? null : valueString;
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexRequestTests.cs]: Remove line 720

File: tests/MongoDB.Driver.Tests/MongoCollectionImplTests.cs [hunk 3]
```diff
@@ -1838,6 +1838,7 @@ private class ClassForDistinctWithArrayField
             [Values(null, 1, 2)] int? commitQuorumW,
             [Values(null, -1, 0, 42, 9000)] int? milliseconds,
             [Values(false, true)] bool usingCreateOneIndexOptions,
+            [Values(false, true)] bool hidden,
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/MongoCollectionImplTests.cs]: Should we test with null also?

File: tests/MongoDB.Driver.Tests/MongoCollectionImplTests.cs [hunk 4]
```diff
@@ -1967,6 +1970,7 @@ private class ClassForDistinctWithArrayField
             [Values(null, 1, 2)] int? commitQuorumW,
             [Values(null, -1, 0, 42, 9000)] int? milliseconds,
             [Values(false, true)] bool usingCreateManyIndexesOptions,
+            [Values(false, true)] bool hidden,
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/MongoCollectionImplTests.cs]: Should we test with null also?

File: tests/MongoDB.Driver.Tests/MongoIndexManagerTests.cs [hunk 5]
```diff
@@ -29,6 +29,7 @@ public class MongoIndexManagerTests
         public void List_should_return_expected_result(
             [Values("{ singleIndex : 1 }", "{ compoundIndex1 : 1, compoundIndex2 : 1 }")] string key,
             [Values(false, true)] bool unique,
+            [Values(false, true)] bool hidden,
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/MongoIndexManagerTests.cs]: Should we test with null also?

File: tests/MongoDB.Driver.Tests/MongoIndexManagerTests.cs [hunk 6]
```diff
@@ -53,9 +64,9 @@ public class MongoIndexManagerTests
                 indexes.Count.Should().Be(2);
                 AssertIndex(collection.CollectionNamespace, indexes[0], "_id_");
                 var indexName = IndexNameHelper.GetIndexName(indexKeyDocument);
-                AssertIndex(collection.CollectionNamespace, indexes[1], indexName, expectedUnique: unique);
+                AssertIndex(collection.CollectionNamespace, indexes[1], indexName, expectedUnique: unique, expectedHidden: hidden);
 
-                void AssertIndex(CollectionNamespace collectionNamespace, BsonDocument index, string expectedName, bool expectedUnique = false)
+                void AssertIndex(CollectionNamespace collectionNamespace, BsonDocument index, string expectedName, bool expectedUnique = false, bool expectedHidden = false)
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/MongoIndexManagerTests.cs]: Should local function be moved to the end of the enclosing function?

**Review by rstam — CHANGES_REQUESTED** [review]
Overall looks good. Minor changes requested.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexRequestTests.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexRequestTests.cs]: done

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexRequestTests.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexRequestTests.cs]: done

File: tests/MongoDB.Driver.Tests/MongoCollectionImplTests.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/MongoCollectionImplTests.cs]: done

File: tests/MongoDB.Driver.Tests/MongoCollectionImplTests.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/MongoCollectionImplTests.cs]: done

File: tests/MongoDB.Driver.Tests/MongoIndexManagerTests.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/MongoIndexManagerTests.cs]: done

File: tests/MongoDB.Driver.Tests/MongoIndexManagerTests.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/MongoIndexManagerTests.cs]: done

**Review by DmitryLukyanov — COMMENTED** [review]
https://evergreen.mongodb.com/version/5f5fd3cb9ccd4e5c346e36e4

ready for review

**Review by rstam — APPROVED** [review]
LGTM

**Review by MikalaiMazurenka — APPROVED** [review]
LGTM!

**Pull request closed without merging**